pkg/scheduler/internal/cache/cache.go

```diff
@@ -230,28 +253,66 @@ func (cache *schedulerCache) UpdateNodeInfoSnapshot(nodeSnapshot *nodeinfosnapsh
 	}
 
 	if len(nodeSnapshot.NodeInfoMap) > len(cache.nodes) {
-		for name := range nodeSnapshot.NodeInfoMap {
-			if _, ok := cache.nodes[name]; !ok {
-				delete(nodeSnapshot.NodeInfoMap, name)
-			}
-		}
+		cache.removeDeletedNodesFromSnapshot(nodeSnapshot)
+		updateAllLists = true
 	}
 
-	// Take a snapshot of the nodes order in the tree
+	if updateAllLists || updateNodesHavePodsWithAffinity {
+		cache.updateNodeInfoSnapshotList(nodeSnapshot, updateAllLists)
+	}
+
+	if len(nodeSnapshot.NodeInfoList) != len(nodeSnapshot.NodeInfoMap) {
+		err := fmt.Errorf("snapshot state is not consistent, length of NodeInfoList=%v not equal to length of NodeInfoMap=%v "+
+			"length of nodes in cache=%v, length of nodes in tree=%v"+
+			", trying to recover",
+			len(nodeSnapshot.NodeInfoList), len(nodeSnapshot.NodeInfoMap),
+			len(cache.nodes), cache.nodeTree.numNodes)
+		// We will try to recover by re-creating the lists for the next scheduling cycle, but still return an
+		// error to surface the problem, the error will likely cause a failure to the current scheduling cycle.
+		cache.updateNodeInfoSnapshotList(nodeSnapshot, true)
```

Comment:
Is this a guaranteed recovery? Or should we recreate the snapshot entirely?

Reply:
from the lists perspective, the map is the source of truth, so I think this is enough for now (or at least for the purpose of what we added). I don't think we have a function to recreate the map.

Reply:
I am afraid if there is a possibility to run into infinite error loop. I think it's safer to recreate the snapshot entirely. Maybe I am just over thinking.

Reply:
This PR does not impact the snapshot map at all, so if we want to recreate it, then we need a better test for checking when it is consistent and when it is not.

Reply:
Yeah I know this PR doesn't change the map. I am OK with the current version.

This is a tricky problem and I really don't know if it could happen or not, and what's the best approach to handle it if it happens. The best approach I can think of is either to recover properly or to panic..